File: src/api/routes/query.py
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

import src.core  # noqa: F401 - triggers strategy registration
from src.api.dependencies import get_session_factory, resolve_strategy
from src.config.settings import settings
from src.core.base_strategy import BaseStrategy
from src.core.models import QueryContext

router = APIRouter()


class QueryRequest(BaseModel):
    patient_id: str = Field(..., min_length=1)
    query: str = Field(..., min_length=1)
    strategy: str | None = None
    model: str | None = None


class QueryResponse(BaseModel):
    response: str
    resource_ids: list[str]
    model_used: str
    strategy_used: str
    tokens_in: int
    tokens_out: int
    latency_ms: float
    resource_types: list[str]
    error: str | None = None

    @classmethod
    def from_result(cls, result) -> "QueryResponse":
        return cls(
            response=result.response_text,
            resource_ids=result.resource_ids,
            model_used=result.model_used,
            strategy_used=result.strategy_used,
            tokens_in=result.tokens_in,
            tokens_out=result.tokens_out,
            latency_ms=result.latency_ms,
            resource_types=getattr(result, "resource_types", []) or [],
            error=result.error,
        )


@router.post("/query", response_model=QueryResponse)
async def run_query(
    req: QueryRequest,
    session_factory=Depends(get_session_factory),
):
    strategy_name = req.strategy or settings.DEFAULT_STRATEGY
    model_name = req.model or settings.DEFAULT_MODEL

    try:
        strategy: BaseStrategy = resolve_strategy(strategy_name, session_factory, model_name)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))

    context = QueryContext(
        patient_id=req.patient_id,
        query_text=req.query,
        strategy_name=strategy_name,
        model_name=model_name,
    )
    result = await strategy.execute(context)
    return QueryResponse.from_result(result)

File: ts-version/src/api/routes/query.ts
import { Router, Request, Response } from 'express';
import { QueryRequestSchema, QueryResponseSchema, QueryContextSchema } from '../../core/models';
import { resolveStrategy } from '../dependencies';
import { logger } from '../../config/logging';
import { getDbPool } from '@/db/session';
import { config } from '@/config/settings';
import { getPatientIdFromHeaders } from '../utils/patientId';

const router = Router();

/**
 * @swagger
 * components:
 *   schemas:
 *     QueryRequest:
 *       type: object
 *       required:
 *         - patientId
 *         - query
 *       properties:
 *         patientId:
 *           type: string
 *           description: Patient ID for the query
 *           example: "2efabc76-892a-b1cf-f47b-4c046d7b197d"
 *         query:
 *           type: string
 *           description: Natural language health query
 *           example: "What medications is this patient taking?"
 *         strategy:
 *           type: string
 *           enum: [langgraph]
 *           description: AI strategy to use for processing
 *           default: "langgraph"
 *         model:
 *           type: string
 *           description: AI model to use
 *           default: "gemini-3-flash-preview"
 *     
 *     QueryResponse:
 *       type: object
 *       properties:
 *         response:
 *           type: string
 *           description: AI-generated response to the query
 *         resourceIds:
 *           type: array
 *           items:
 *             type: string
 *           description: List of FHIR resource IDs referenced in the response
 *         modelUsed:
 *           type: string
 *           description: The AI model that was used
 *         strategyUsed:
 *           type: string
 *           description: The strategy that was used
 *         tokensIn:
 *           type: number
 *           description: Number of input tokens used
 *         tokensOut:
 *           type: number
 *           description: Number of output tokens generated
 *         latencyMs:
 *           type: number
 *           description: Processing latency in milliseconds
 *         error:
 *           type: string
 *           description: Error message if processing failed
 *     
 *     ErrorResponse:
 *       type: object
 *       properties:
 *         error:
 *           type: string
 *           description: Error type
 *         message:
 *           type: string
 *           description: Detailed error message
 */

/**
 * @swagger
 * /api/fhir/query:
 *   post:
 *     summary: Process a health query using AI
 *     description: Submits a natural language health query for a specific patient and returns an AI-generated response with relevant FHIR resources
 *     tags: [Query]
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             $ref: '#/components/schemas/QueryRequest'
 *     responses:
 *       200:
 *         description: Query processed successfully
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/QueryResponse'
 *       400:
 *         description: Invalid request
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/ErrorResponse'
 *       500:
 *         description: Processing failed
 *         content:
 *           application/json:
 *             schema:
 *               $ref: '#/components/schemas/ErrorResponse'
 */
// Dependency injection function (match Python: Depends(get_session_factory))
function getSessionFactory() {
  return getDbPool();
}

router.post('/query', async (req: Request, res: Response) => {
  try {
    // Validate request
    const validatedQuery = QueryRequestSchema.parse(req.body);
    
    // Get patientId from request body or headers
    let patientId = validatedQuery.patientId;
    if (!patientId) {
      patientId = await getPatientIdFromHeaders(req);
    }
    
    const strategyName = validatedQuery.strategy || config.DEFAULT_STRATEGY;
    const modelName = validatedQuery.model || config.DEFAULT_MODEL;

    // Get session factory (match Python: session_factory=Depends(get_session_factory))
    const sessionFactory = getSessionFactory();

    try {
      // Resolve strategy (match Python: resolve_strategy(strategy_name, session_factory, model_name))
      const strategy = resolveStrategy(strategyName, sessionFactory, modelName);
      
      // Execute query (match Python: QueryContext)
      const context = QueryContextSchema.parse({
        patientId,
        queryText: validatedQuery.query,
        strategyName,
        modelName,
      });

      const result = await strategy.execute(context);
      
      // Return response (match Python: QueryResponse.from_result(result))
      return res.json(result);
    } catch (error) {
      // Match Python: HTTPException(status_code=400, detail=str(e))
      const errorMessage = error instanceof Error ? error.message : String(error);
      return res.status(400).json({
        error: 'Query processing failed',
        message: errorMessage,
      });
    }
  } catch (error) {
    // Request validation error
    const errorMessage = error instanceof Error ? error.message : String(error);
    return res.status(400).json({
      error: 'Invalid request',
      message: errorMessage,
    });
  }
});

export { router as queryRouter };
